### ISSUE TITLE
Much higher GPU memory usage compared to LLaMA-Factory

### ISSUE THREAD

**Issue by baicenxiao** [2U1/Qwen2-VL-Finetune#122]
Thanks for providing this useful repo! The training script works, but I observed a much higher GPU memory usage when using sft to tune (full) Qwen2.5-vl-3B-Instruct, compared to the same setting/configuration but trained with LLaMA-Factory. 

I used 4 A10g GPUs (4 x 24G vRAM) with the default zero3-offload config from this repo. Below is the script I modified from the repo:

```
#!/bin/bash

# You can use 2B instead of 7B
# MODEL_NAME="Qwen/Qwen2-VL-7B-Instruct"
# MODEL_NAME="Qwen/Qwen2-VL-2B-Instruct"
MODEL_NAME="Qwen/Qwen2.5-VL-3B-Instruct"
# MODEL_NAME="Qwen/Qwen2.5-VL-7B-Instruct"

GLOBAL_BATCH_SIZE=16
BATCH_PER_DEVICE=1
NUM_DEVICES=4
GRAD_ACCUM_STEPS=$((GLOBAL_BATCH_SIZE / (BATCH_PER_DEVICE * NUM_DEVICES)))

export PYTHONPATH=src:$PYTHONPATH

deepspeed src/train/train_sft.py \
    --use_liger False \
    --deepspeed scripts/zero3_offload.json \
    --model_id $MODEL_NAME \
    --data_path /OCR_subset.json \
    --image_folder /OCR_subset_images \
    --remove_unused_columns False \
    --freeze_vision_tower False \
    --freeze_llm False \
    --freeze_merger False \
    --bf16 True \
    --fp16 False \
    --disable_flash_attn2 False \
    --output_dir output/ \
    --num_train_epochs 3 \
    --per_device_train_batch_size $BATCH_PER_DEVICE \
    --gradient_accumulation_steps $GRAD_ACCUM_STEPS \
    --image_min_pixels $((512 * 28 * 28)) \
    --image_max_pixels $((1280 * 28 * 28)) \
    --learning_rate 1e-5 \
    --merger_lr 1e-5 \
    --vision_lr 2e-6 \
    --weight_decay 0.1 \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --gradient_checkpointing True \
    --report_to none \
    --lazy_preprocess True \
    --save_strategy "steps" \
    --save_steps 200 \
    --save_total_limit 3 \
    --dataloader_num_workers 4 \
    --max_seq_length 8192
```
Below is GPU usage from nvidia-smi, where you can see the vRAM of GPU can easily go above 20GB:

<img width="603" alt="Image" src="https://github.com/user-attachments/assets/de6d7b78-fba2-44dd-a922-b9d94602bee6" />

Below is the config I used with LLaMA-Factory:
```
### model
model_name_or_path: /sensei-fs/users/bxiao/Qwen2-VL-finetune-LatexOCR/Qwen/Qwen2.5-VL-3B-Instruct
image_max_pixels: 1003520
image_min_pixels: 401408

video_max_pixels: 16384
trust_remote_code: true

### method
stage: sft
do_train: true
finetuning_type: full
freeze_vision_tower: false
freeze_multi_modal_projector: false
freeze_language_model: false
deepspeed: examples/deepspeed/ds_z3_offload_config.json

### dataset
dataset: ocr_train
template: qwen2_vl
cutoff_len: 8192
max_samples: 1000
overwrite_cache: true
preprocessing_num_workers: 16
dataloader_num_workers: 4

### output
output_dir: saves/qwen2_5vl-7b/full/sft
logging_steps: 10
save_steps: 500
plot_loss: true
overwrite_output_dir: true
save_only_model: false
report_to: none  # choices: [none, wandb, tensorboard, swanlab, mlflow]

### train
per_device_train_batch_size: 1
gradient_accumulation_steps: 2
learning_rate: 1.0e-5
num_train_epochs: 3.0
lr_scheduler_type: cosine
warmup_ratio: 0.1
bf16: true
ddp_timeout: 180000000
resume_from_checkpoint: null

### eval
# val_size: 0.1
# per_device_eval_batch_size: 1
# eval_strategy: steps
# eval_steps: 500

```

Below is GPU usage with LLaMA-Factory, where GPU vRAM usages are generally below 13GB. I used the same training data.

<img width="603" alt="Image" src="https://github.com/user-attachments/assets/dcfe8809-6c79-4616-9da4-98d22f2a6ef4" />

Could you please check this issue? Any hints/ ideas would be appreciated. 

Here is Google drive [link](https://drive.google.com/file/d/1IZgjdAz8BIzLB62jTsBd8iDv6J7gZwK2/view?usp=sharing) to a complete subset of the data I use, for the purpose of reproducing the issue.

**Comment by 2U1** [comment]
Thanks for letting me know.

First, please make sure that `use_liger` is set to True when you train with this repository.
Second, the current version of my repo doesn’t yet support the `max_seq_length` argument. I plan to add dynamic truncation, but I’m not sure of the best way to implement it—if the limit is too short, it might cut off part of the user’s input.

**Comment by 2U1** [comment]
`"overlap_comm": true,` also this could cause a bit more memory.

**Comment by baicenxiao** [comment]
Another question is, I tried to use Huggingface trainer to directly finetune `Qwen2.5-vl-3B-Instruct`. Using the same zero3-offload config, gradient checkpointing and liger-kernel, I observed much higher vRAM cost compared to finetuning using your repo:

<img width="605" alt="Image" src="https://github.com/user-attachments/assets/fc1dcfbb-35e5-4d0c-8dfc-67a5d4009393" />

Could you please give some guidance or idea which part of your code contributes to the vRAM saving? Really want to learn more.


FYR, here is the outline of code I use to finetune with Huggingface trainer:
```
local_model_path = "./Qwen/Qwen2.5-VL-3B-Instruct"
train_dataset_json_path = "ocr_train_subset.json"
output_dir = "./output/Qwen2-5-VL-3B-ds_stage3" 
MAX_LENGTH = 8192
DEEPSPEED_CONFIG = "../LLaMA-Factory/examples/deepspeed/ds_z3_offload_config.json"

# Add liger kernel
apply_liger_kernel_to_qwen2_5_vl()

# Load model
origin_model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    local_model_path,
    torch_dtype=torch.bfloat16, # Keep bfloat16 for consistency with deepspeed config
    trust_remote_code=True,
    use_cache=False,
    attn_implementation="flash_attention_2"
)

origin_model.config.use_cache = False
origin_model.gradient_checkpointing_enable()


# # Process datasets
train_ds = Dataset.from_json(train_dataset_json_path)
train_dataset = train_ds.with_transform(process_func_batch) # <<< CHANGE: Use with_transform


# Configure TrainingArguments 
args = TrainingArguments(
    output_dir=output_dir,
    per_device_train_batch_size=1, 
    gradient_accumulation_steps=4, 
    num_train_epochs=3,
    save_strategy="steps", 
    save_steps=20,
    logging_steps=10,
    logging_first_step=True, 
    learning_rate=1e-5,
    save_on_each_node=False, 
    gradient_checkpointing=True, # Already enabled on model, but set here too for clarity
    report_to="none", 
    remove_unused_columns=False, 
    bf16=True, 
    deepspeed=DEEPSPEED_CONFIG, # Zero stage 3 with offload
    dataloader_num_workers=NUM_CPU
)

# Configure Trainer
trainer = Trainer(
    model=origin_model, # Trainer handles DeepSpeed wrapping
    args=args,
    train_dataset=train_dataset,
    data_collator=DataCollatorForSeq2Seq(tokenizer=tokenizer, padding=True), # Pad to longest in batch
    callbacks=callbacks,
)

# Start training
train_result = trainer.train()
```

**Comment by 2U1** [comment]
I'm not really sure what is the difference. I think llama-factory is based on the code from official qwen-vl repo, but that is not quite different from mine except for the data packing part. 

Could you log the peak usage of the vram?

**Comment by baicenxiao** [comment]
Below are monitored vRAM usage within 5 minutes from training.

llama-factory **with** liger kernel:
```
GPU 0: Peak = 5961.75 MB, 
GPU 1: Peak = 5839.75 MB, 
GPU 2: Peak = 5899.75 MB, 
**GPU 3: Peak = 5961.75 MB,** 
```

2U1 framework **with** liger kernel:
```
GPU 0: Peak = 6269.75 MB
GPU 1: Peak = 6079.75 MB
**GPU 2: Peak = 8323.75 MB**
GPU 3: Peak = 7661.75 MB
```

llama-factory **without** liger kernel:
```
GPU 0: Peak = 11215.75 MB, 
**GPU 1: Peak = 13617.75 MB**
GPU 2: Peak = 11535.75 MB
GPU 3: Peak = 13217.75 MB
```

2U1 framework **without** liger kernel:
```
GPU 0: Peak = 17533.75 MB
**GPU 1: Peak = 21887.75 MB**
GPU 2: Peak = 21037.75 MB
GPU 3: Peak = 16783.75 MB
```

**Comment by baicenxiao** [comment]
@2U1 I just tried setting `{'use_reentrant':True}` in Huggingface trainer, but I did not notice any vRAM usage improvement. Wondering if there is any other reason leading to better vRAM in this 2UI framework.

**Comment by 2U1** [comment]
@baicenxiao I'm not sure why llama-factory uses less vram. I've looked into the code but it's not so different from the code I've made. I've checked the default optimizer and other settings that maybe it could be the reason but I couldn't find it.

Also, for the huggingface trainer, using reentrant should use less memory. It shows difference in the code when I was testing. Or maybe the resizing isn't working well in the code?

**Comment by 2U1** [comment]
I've made a additional monkey patching in the forward fuction, and it will run with much less memory and much faster speed.
The code will be updated when the test is finished (Later today).
